test/headers.js

```diff
@@ -0,0 +1,240 @@
+
+	const headers = await ky.get(server.url).json();
+	t.is(headers.host, `localhost:${server.port}`);
+});
+
+test('transforms names to lowercase', async t => {
+	const server = await createTestServer();
+	server.get('/', echoHeaders);
+
+	const headers = await ky(server.url, {
+		headers: {
+			'ACCEPT-ENCODING': 'identity'
+		},
+		responseType: 'json'
+	}).json();
+	t.is(headers['accept-encoding'], 'identity');
+});
+
+test('setting `content-length` to 0', async t => {
+	const server = await createTestServer();
+	server.post('/', echoHeaders);
+
+	const headers = await ky
+		.post(server.url, {
+			headers: {
+				'content-length': '0'
+			},
+			body: 'sup'
+		})
+		.json();
+
+	t.is(headers['content-length'], '3');
+});
+
+test('sets `content-length` to `0` when requesting PUT with empty body', async t => {
+	const server = await createTestServer();
+	server.put('/', echoHeaders);
+
+	const headers = await ky.put(server.url).json();
+
+	t.is(headers['content-length'], '0');
+});
+
+test('form-data manual `content-type` header', async t => {
+	const server = await createTestServer();
+	server.post('/', echoHeaders);
+
+	const form = new FormData();
+	form.append('a', 'b');
+	const headers = await ky
+		.post(server.url, {
+			headers: {
+				'content-type': 'custom'
+			},
+			body: form
+		})
+		.json();
+
+	t.is(headers['content-type'], 'custom');
+});
+
+test('form-data automatic `content-type` header', async t => {
+	const server = await createTestServer();
+	server.post('/', echoHeaders);
+
+	const form = new FormData();
+	form.append('a', 'b');
+	const headers = await ky.post(server.url, {
+		body: form
+	}).json();
+
+	t.is(headers['content-type'], `multipart/form-data;boundary=${form.getBoundary()}`);
+});
+
+test('form-data sets `content-length` header', async t => {
+	const server = await createTestServer();
+	server.post('/', echoHeaders);
+
+	const form = new FormData();
+	form.append('a', 'b');
+	const headers = await ky.post(server.url, {body: form}).json();
+
+	t.is(headers['content-length'], '157');
+});
+
+test('buffer as `options.body` sets `content-length` header', async t => {
+	const server = await createTestServer();
+	server.post('/', echoHeaders);
+
+	const buffer = Buffer.from('unicorn');
+	const headers = await ky.post(server.url, {
+		body: buffer
+	}).json();
+
+	t.is(Number(headers['content-length']), buffer.length);
+});
+
+test('removes undefined value headers', async t => {
+	const server = await createTestServer();
+	server.get('/', echoHeaders);
+
+	const headers = await ky.get(server.url, {
+		headers: {
+			'user-agent': undefined
+		}
+	}).json();
+
+	t.is(headers['user-agent'], 'undefined');
+});
+
+test.failing('non-existent headers set to undefined are omitted', async t => {
+	const server = await createTestServer();
+	server.get('/', echoHeaders);
+
+	const headers = await ky.get(server.url, {
+		headers: {
+			blah: undefined
+		}
+	}).json();
+
+	t.false(Reflect.has(headers, 'blah'));
```

Comment:
nit: personally, I prefer using `'blah' in headers` over `Reflect.has(headers, 'blah')`. Not sure how @sindresorhus feels about it, though.

Reply:
<details>
<summary>I did some benchmarks</summary>

![image](https://user-images.githubusercontent.com/36894700/70862435-c9102a00-1f3c-11ea-92eb-8a0fddc0dd76.png)

</details>

Reply:
Micro-benchmarks are not that useful. I don't feel strongly either way other than being consistent. 

Is `Reflect.has` currently enforced? If not, maybe worth opening an issue on https://github.com/sindresorhus/eslint-plugin-unicorn about adding a rule for enforcing either of the options?

I think I started using `Reflect.has` initially as I naively thought it was stricter, but it's the same, so maybe `in` is just better.